src/targets/gpu/jit/mlir.cpp

```diff
@@ -21,39 +21,196 @@
+        }
+        else if(i->name() == "@literal")
+        {
+            ins_map[i] = m.add_literal(i->get_literal());
+        }
+        else if(i->name() == "multibroadcast" and i->inputs().front()->name() == "@literal")
+        {
+            ins_map[i] = ins_map[i->inputs().front()];
+        }
+        else
+        {
+            std::vector<instruction_ref> inputs;
+            inputs.resize(i->inputs().size());
+            std::transform(i->inputs().begin(),
+                           i->inputs().end(),
+                           inputs.begin(),
+                           [&](const auto input) { return ins_map[input]; });
+            if(i->name() == "@return")
+            {
+                m.add_return(inputs);
+            }
+            else
+            {
+                ins_map[i] = m.add_instruction(i->get_operator(), inputs);
+            }
+        }
+    }
+    return m;
+}
+
 struct mlir_compiler : compiler<mlir_compiler>
 {
     std::vector<std::string> names() const { return {"gpu::mlir_op"}; }
 
     operation compile_op(context&, const std::vector<shape>&, const value&) const { return {}; }
 
     compiler_replace
-    compile(const context& ctx, instruction_ref ins, const operation&, const value& solution) const
+    compile(context& ctx, instruction_ref ins, const operation&, const value& solution) const
     {
         auto* smod = ins->module_inputs().front();
         assert(smod->get_parameter_names().size() == ins->inputs().size() - 1);
-        return insert(compile_mlir(ctx, *smod, ins->inputs(), solution));
+        auto gemm_ins = std::find_if(smod->begin(), smod->end(), [&](const auto& i) {
+            return i.name() == "dot" or i.name() == "quant_dot";
+        });
+        // check if (a) module is fused (b) contains a dot instruction and (c) perfConfig can not
+        // allow fused module
+        if(gemm_ins != smod->end() and std::distance(gemm_ins, smod->end()) > 2 and
+           not is_module_fusible(*smod, solution))
+        {
+            auto input_args = ins->inputs();
+            input_args.pop_back();
+            auto mod_splits      = smod->split(input_args, {gemm_ins});
+            auto dot_mlir_inputs = to_shapes(mod_splits[0].inputs);
+            dot_mlir_inputs.push_back(mod_splits[0].mod.get_output_shapes().front());
+            mlir_code_object cop1 = compile_mlir(ctx, mod_splits[0].mod, dot_mlir_inputs, solution);
+            auto pw_inputs        = mod_splits[1].inputs;
+            auto dot_ins_idx      = std::distance(
+                std::find(pw_inputs.begin(), pw_inputs.end(), gemm_ins), pw_inputs.begin());
+            auto pw_shapes         = to_shapes(mod_splits[1].inputs);
+            pw_shapes[dot_ins_idx] = cop1.cop.output;
+            pw_shapes.push_back(mod_splits[1].mod.get_output_shapes().front());
+            assert(pw_shapes.back() == ins->get_shape());
+            auto pw_mod                        = create_pointwise_module(&mod_splits[1].mod);
```

Comment:
Why is `create_pointwise_module` called? Why not pass `mod_splits[1].mod` directly to `compile_pointwise`?

Reply:
`mod_splits[1].mod` have multibroadcasts instructions which need to be eliminated otherwise compile_pointwise fails.